### PR TITLE
make hdf5 optional

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Description

make HDF5 optional

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Comment by timosachsenberg** [comment]
Is there a better way so that these if / else are not spread out over so many files?
E.g., a more "object-oriented" way of writing cmake scripts?

will add the if else for HDF5::HDF5 ...

File: src/tests/class_tests/openms/CMakeLists.txt [hunk 1]
```diff
@@ -91,7 +91,10 @@ target_link_libraries(Base64_test ZLIB::ZLIB)
 target_link_libraries(LPWrapper_test CoinOR::CoinOR)
 target_link_libraries(SpectraSTSimilarityScore_test Eigen3::Eigen)
 target_link_libraries(BinnedSpectrum_test Eigen3::Eigen)
-target_link_libraries(HDF5_test HDF5::HDF5)
+
+if (WITH_HDF5)
+  target_link_libraries(HDF5_test HDF5::HDF5)
```
Comment (jpfeuffer) [inline src/tests/class_tests/openms/CMakeLists.txt:96]: Maybe the same needs to be done for the OpenMS lib?

Comment (timosachsenberg) [reply src/tests/class_tests/openms/CMakeLists.txt:96]: yeah I will add it...

**Comment by jpfeuffer** [comment]
In theory, you could modify the targets at the very end. But then you will have if cases for each target in the contrib related code.

If with_gui, ....

**Comment by timosachsenberg** [comment]
if you know: is it possible or a good idea to turn
```
set(OPENMS_DEP_PRIVATE_LIBRARIES GTE Quadtree ${LPTARGET} Eigen3::Eigen HDF5::HDF5 Boost::iostreams Boost::date_time Boost::boost Boost::regex BZip2::BZip2 ZLIB::ZLIB SQLiteCpp)
```
into:
```
set(OPENMS_DEP_PRIVATE_LIBRARIES GTE Quadtree ${LPTARGET} Eigen3::Eigen $<$<BOOL:${WITH_HDF5}>:HDF5::HDF5> Boost::iostreams Boost::date_time Boost::boost Boost::regex BZip2::BZip2 ZLIB::ZLIB SQLiteCpp)
```

**Comment by jpfeuffer** [comment]
Yes should work. Do you like it better than an ifcase?

File: src/openms/CMakeLists.txt [hunk 2]
```diff
@@ -88,7 +88,17 @@ include (${PROJECT_SOURCE_DIR}/includes.cmake)
 # all the dependency libraries are linked into libOpenMS.so
 set(OPENMS_DEP_LIBRARIES Evergreen LibSVM::LibSVM XercesC::XercesC Qt5::Core Qt5::Network)
 
-set(OPENMS_DEP_PRIVATE_LIBRARIES GTE Quadtree ${LPTARGET} Eigen3::Eigen HDF5::HDF5 Boost::iostreams Boost::date_time Boost::boost Boost::regex BZip2::BZip2 ZLIB::ZLIB SQLiteCpp)
+set(OPENMS_DEP_PRIVATE_LIBRARIES 
+  GTE Quadtree ${LPTARGET} 
+  Eigen3::Eigen
+  $<$<BOOL:${WITH_HDF5}>:HDF5::HDF5>
```
Comment (timosachsenberg) [inline src/openms/CMakeLists.txt:94]: e.g., could also move it to the bottom and add a comment.
Think that is needed?

File: src/openms/source/FORMAT/sources.cmake [hunk 3]
```diff
@@ -101,6 +100,10 @@ XTandemXMLFile.cpp
 ZlibCompression.cpp
 )
 
+if (WITH_HDF5)
+  list(APPEND sources_list HDF5Connector.cpp)
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/sources.cmake:104]: could probably also be moved into the list...

Comment (timosachsenberg) [reply src/openms/source/FORMAT/sources.cmake:104]: Ah nope doesn’t work

**Review by jpfeuffer — APPROVED** [review]

**Pull request merged**